source/languageDetection.py

```diff
@@ -0,0 +1,284 @@
+		return  languageHandler.getLanguage()
+	# default language is not applicable for this script, so look up in the scriptIDToLangID
+	return scriptIDToLangID.get (scriptName )
+
+def getLanguageDescription(language ):
+	desc=languageHandler.getLanguageDescription(language )
+	label="%s, %s"%(desc,language ) if desc else language 
+	return label
+
+def getScriptIDFromLangID(langID ):
+	# Strip the dialect (if any).
+	return langIDToScriptID.get (langID.split("_", 1)[0])
+
+class ScriptChangeCommand(SpeechCommand):
+	"""A command to switch the script during script detection ."""
+
+	def __init__(self, scriptCode):
+		"""
+		@param scriptCode: the script identifier
+		@type scriptCode: int
+		"""
+		self.scriptCode =scriptCode 
+
+	def __repr__(self):
+		return "ScriptChangeCommand (%r)"%self.scriptCode
+
+def detectScript(text):
+	"""splits a string if there are multiple scripts in it
+	@param text: the text string
+	@type string
+	@return: sequence of script commands and text
+	@rtype: list"""
+	unicodeSequence = []
+	outerScripts  = []
+	currentScript = getScriptCode(  text[0])
+	oldScript = currentScript
+	unicodeSequence.append(ScriptChangeCommand(currentScript)) 
+	beginIndex = 0
+	for index in xrange( len(text) ) :
+		currentScript = getScriptCode( text[index] ) 
+		# handling script property for paired punctuation: see UAX 24 section 5.1
+		if unicodedata.category(text[index] ) == "Ps":
+			outerScripts.append(oldScript ) 
+		elif unicodedata.category(text[index] ) == "Pe" and len(outerScripts) > 0:
+			currentScript = outerScripts.pop()
+
+		# don't change script for combining marks as per UAX 24 section 5.2
+		if unicodedata.category(text[index] ) in ["Mn" , "Mc", "Me"]: 
+			continue
+
+		# don't change scripts for Common, inherited and Unknown characters 
+		if currentScript in ["Common" , "Inherited", "Unknown"]: 
+			continue
+
+		if currentScript != oldScript:
+			newText = text[beginIndex:index] 
+			unicodeSequence.append( newText )
+			beginIndex= index
+			unicodeSequence.append(ScriptChangeCommand(currentScript)) 
+		oldScript = currentScript
+
+	unicodeSequence.append( text[beginIndex:] )
+	return unicodeSequence
+
+def detectLanguage(text, defaultLanguage =None):
+	"""splits a string if there are multiple languages in it. uses detectScript
+	@param text: the text string
+	@type text: string
+	@param defaultLanguage: The default language for NVDA 
+	@type preferredLanguage: string or None
+	@return: sequence of language commands and text
+	@rtype: list"""
+	sequenceWithLanguage= []
+	if config.conf["languageDetection"]["disableScriptDetection"]:  
+		sequenceWithLanguage.append( text)
+		return sequenceWithLanguage
+
+
+	tempSequence = detectScript(text)
+	scriptCode = ""
+	for item in tempSequence: 
+		if isinstance(item,ScriptChangeCommand):
+			scriptCode = item.scriptCode
+		else:
+			log.debugWarning(u"script: {} for text {} ".format( scriptCode , unicode(item) ) )
```

Comment:
Should this be here? What does this log message tell us?

Reply:
Removed